_source/_posts/2021-12-03-antipatterns-secrets-java.adoc

```diff
@@ -0,0 +1,149 @@
+[%collapsible]
+=====
+image::{% asset_path 'blog/antipatterns-secrets-java/facepalm.gif' %}[alt=Picard facepalm from www.reactiongifs.com/picard-facepalm/,width=400,align=left]
+=====
+====
+
+
+== 4. Restrict access to secrets
+
+It's hard to keep secrets, well... secret. Do you know who has access to each secret used by your application? Secrets kept in environment variables can be viewed by anyone who has access to the running process on the system `ps e -ww -p <pid>`, or access to the Docker demean `docker inspect <container>`.  Even "hidden" secrets used by your build server are https://www.theserverside.com/blog/Coffee-Talk-Java-News-Stories-and-Opinions/GitHub-Actions-Secrets-Example-Token-Tutorial[easy to leak]. Most popular CI services will filter build logs of passwords, but it's easy to work around this. For example, this GitHub Actions script adds a space between each character and bypasses the filtering.
+
+{% raw %}
+[source,yaml]
+----
+steps:
+- name: Not So Hidden Secret
+  run: "echo ${{secrets.TEST_SECRET}} | sed 's/./& /g'"
+
+# output: C a n   y o u   s t i l l   r e a d   t h i s   t e x t ?
+----
+{% endraw %}
+
+WARNING: I'm not suggesting you do this; doing so may get you fired. Just be aware that it can happen and treat your secrets accordingly.
+
+
+== 5. Building your own crypto
+
+As the old saying goes, "Don't roll your own crypto." You are probably aware that you shouldn't implement crypto algorithms yourself, but this saying means much more than that. How the output is used and the protocol involved is just as important as the actual algorithm itself. Different algorithms have been designed for different use cases; many protocols or formats are designed with these in mind.
+
+Instead of cobbling together lower-level primitives, use a high-level library.  This suggestion may be obvious for protocols like TLS/HTTPS, but it also applies to _simple_ formats; a lot can go wrong when a link:/blog/2020/12/21/beginners-guide-to-jwt#problems[JWT is parsed without a quality library].
+
+
+== Bonus: Be ready to rotate your secrets
+
+If history has taught us anything about secrets, it's that they will leak. In the last year there have been a few notable cases of secrets being exposed: https://travis-ci.community/t/security-bulletin/12081[Travis-CI] and https://about.codecov.io/security-update/[Codecov]. After these issues were disclosed, teams scrambled to update secrets that may have been affected, often resulting in broken builds.
+
+IMPORTANT: When people leave your project, any secrets they had access to must be updated.
+
+Be aware of how your project uses secrets, and practice rotating them before the next leak or personal change.
+
+
+== Learn more about application security
+
+I've covered a few common problems with managing secrets for Java applications in this post. But this barely scratches the surface of application security and secret management. Check out these other posts to learn more:
```

Comment:
```suggestion
In this post, I've described a few common anti-patterns to avoid when managing secrets for Java applications. But this list of five barely scratches the surface of application security and secret management. Check out these other posts to learn more:
```